@Factory with dataProvider changes order of iterations

krmahadevan opened the pull request:
Closes #799

Fixes #799 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

krmahadevan commented:
ping @juherr

juherr commented on src/main/java/org/testng/internal/Systematiser.java:
```diff
@@ -0,0 +1,86 @@
+
+import java.util.Comparator;
+
+public class Systematiser {
```
I'm not a huge fan of the name but I've nothing better to propose :)

juherr commented on src/main/java/org/testng/internal/Systematiser.java:
```diff
@@ -0,0 +1,86 @@
+public class Systematiser {
+
+    private Systematiser() {
+        //Utility class. Defeat instantiation.
```
The class should be final too.

krmahadevan replied:
fixed

juherr commented on src/main/java/org/testng/internal/Graph.java:
```diff
@@ -121,7 +122,10 @@ public void topologicalSort() {
     // get run close to each other as much as possible
     //
-    Collections.sort(nodes2);
+    Comparator<Node> comparator = Systematiser.getComparator();
```
Could you add the `Comparator<Node<T>>` as constructor param?

krmahadevan replied:
I didnt quite understand this observation @juherr . Can you please elaborate.

juherr replied:
The comparator should be an attribute of the graph and the comparator instantiation logic should be outside.
The attribute description will be: `private final Comparator<Node<T>> comparator` and its usage will be `Collections.sort(nodes2, comparator)`.

krmahadevan replied:
Ok, I understood. But why would the `comparator` instantiation be plugged into a `Graph` instance from outside ? The `comparator` instantiation is completely dependent on just a JVM argument and as such I thought we can have the `comparator` be a field within `Graph` but skip the constructor parameter part.

juherr replied:
JVM argument is just a quick workaround which should not be used in the core.
I don't like it because it is a global variable which is bad by definition.
That's why I think we should manage the feature from a global point of view but offer JVM arg to the users in order to keep API in the same state.
At the same time, if the integration of the feature modifies all the structure, we may use some temporary tricks and think about the core redesign later.

krmahadevan replied:
From what I have looked at, if one were to be resorting to make this available via a configuration instead of a JVM argument, this is going to require quite a few changes. Also if you look at it, a configuration would still be applicable at the entire TestNG instance level, which is kind of what the JVM argument is doing as well. Does that add a bit of clarity to your question ?

juherr replied:
I will try to have a deeper look asap.

juherr commented on src/main/java/org/testng/internal/Systematiser.java:
```diff
@@ -0,0 +1,86 @@
+                break;
+
+            case NONE:
+            default:
```
Instead of null, maybe we can use a default comparator (all params are equals)?

krmahadevan replied:
Fixed.

juherr commented on src/main/java/org/testng/internal/Systematiser.java:
```diff
@@ -0,0 +1,86 @@
+        //Utility class. Defeat instantiation.
+    }
+
+    public static Comparator<Graph.Node> getComparator() {
```
`Comparator<Graph.Node<ITestNGMethod>>`

krmahadevan replied:
We cannot do this because the INSTANCES one doesn't work with a `ITestNGMethod` but merely works with just an `Object`. The `ITestNGMethod` based comparator also if you notice resorts to using the `ITestNGMethod` object **iff** the Node contains an `ITestNGMethod` object.

juherr replied:
I don't catch it because `o1.getObject().toString()` will work with `Node<ITestNGMethod>` too.

juherr replied:
In fact, this point is important IMO.

krmahadevan replied:
@juherr - Ok, but I would like to understand why does `Node` object have to be tied to `ITestNGMethod` ? There are tests for the `Graph` class viz., [GraphTest](https://github.com/cbeust/testng/blob/master/src/test/java/test/GraphTest.java#L21-L27) which attempts at testing `Graph` behavior using Strings. If we basically tie down the `Comparator` to `ITestNGMethod` then this would cause changes to be required in the test as well and by behavior one shouldn't have to need an `ITestNGMethod` object to test the behavior of the `Graph`. That is why I didn't change it.

krmahadevan commented:
@juherr - I have addressed some of the review comments and responded to the ones that I have not fixed. Please help take a look and let me know if this can be approved.

juherr commented on src/test/resources/testng.xml:
```diff
@@ -815,12 +815,19 @@
     </classes>
   </test>
+
+  <test name="RunOrderingTest">
```
Now, we may add a programmatically way to configure the test without system properties.

cbeust commented on src/main/java/org/testng/internal/Systematiser.java:
```diff
@@ -0,0 +1,96 @@
+
+import java.util.Comparator;
+
+public final class Systematiser {
```
We need a Javadoc for this class, especially since its name is so bizarre.

krmahadevan replied:
@cbeust - I have added the required javadocs as well. Can you please see if thats fine ?

krmahadevan replied:
If you have some other name in mind please let me know @cbeust I can rename this. I named this so, keeping in mind it helps organize methods in some fashion. But if the name is odd, I can have it named to something like `ComparatorFactory` or something on those lines
